packages/pds/src/db/message-queue/index.ts

```diff
@@ -0,0 +1,288 @@
+import { RepoStructure } from '@atproto/repo'
+import Database from '..'
+import {
+  AddMember,
+  AddUpvote,
+  CreateNotification,
+  DeleteNotifications,
+  Message,
+  RemoveMember,
+  RemoveUpvote,
+} from './messages'
+import { sql } from 'kysely'
+import { dbLogger as log } from '../../logger'
+import { AuthStore } from '@atproto/auth'
+import * as repoUtil from '../../util/repo'
+import * as schema from '../../lexicon/schemas'
+import { TID } from '@atproto/common'
+import { MessageQueue } from '../types'
+
+type GetAuthStoreFn = (did: string) => AuthStore
+
+export class SqlMessageQueue implements MessageQueue {
+  private cursorExists = false
+
+  constructor(
+    private name: string,
+    private db: Database,
+    private getAuthStore: GetAuthStoreFn,
+  ) {}
+
+  async send(tx: Database, messages: Message | Message[]): Promise<void> {
+    const msgArray = Array.isArray(messages) ? messages : [messages]
+    if (msgArray.length === 0) return
+    const now = new Date().toISOString()
+    const values = msgArray.map((msg) => ({
+      message: JSON.stringify(msg),
+      createdAt: now,
+    }))
+
+    await tx.db.insertInto('message_queue').values(values).execute()
+
+    this.process()
+  }
+
+  private async ensureCursor(): Promise<void> {
+    if (this.cursorExists) return
+    await this.db.db
+      .insertInto('message_queue_cursor')
+      .values({ consumer: this.name, cursor: 1 })
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    this.cursorExists = true
+  }
+
+  async process(): Promise<void> {
+    await this.ensureCursor()
+    const keepGoing = await this.db.transaction(
+      async (dbTxn): Promise<boolean> => {
+        let builder = dbTxn.db
+          .selectFrom('message_queue_cursor as cursor')
+          .innerJoin('message_queue', 'message_queue.id', 'cursor.cursor')
+          .selectAll()
+        if (this.db.dialect !== 'sqlite') {
+          builder = builder.forUpdate()
+        }
+
+        const res = await builder.executeTakeFirst()
+        // all caught up
+        if (!res) return false
+
+        const message: Message = JSON.parse(res.message)
+        await this.handleMessage(dbTxn, message)
+        await dbTxn.db
+          .updateTable('message_queue_cursor')
+          .set({ cursor: sql`cursor + 1` })
```

Comment:
Only question I have about this is if we can assume the incrementing id is guaranteed to not leave any gaps.  Otherwise I think the cursor can get stuck.

Reply:
oh that's a good point. i _think_ so? worth a look 🧐

Reply:
it doesn't guarantee it :notlikethis:

Reply:
kk this should fix it
https://github.com/bluesky-social/atproto/pull/327/commits/e4464814e4f03b50ce7aec963b24ca45803f67cd

(also added an actual SQL LIMIT in the next commit)